Sources/Views/MessageLabel.swift

```diff
@@ -22,417 +22,491 @@
-        textContainer.maximumNumberOfLines = self.numberOfLines
-        textContainer.lineBreakMode = self.lineBreakMode
-        textContainer.size = self.bounds.size
-        return textContainer
-    }()
+// adjustsFontSizeToFitWidth
+// allowsDefaultTighteningForTruncation
+// baselineAdjustment
+// minimumScaleFactor
+// highlightedTextColor
+// isHighlighted
+// shadowColor
+// shadowOffset
 
-    private lazy var textStorage: NSTextStorage = {
-        let textStorage = NSTextStorage()
-        textStorage.addLayoutManager(self.layoutManager)
-        return textStorage
-    }()
-
-    private lazy var rangesForDetectors: [DetectorType: [(NSRange, Any?)]] = [:]
-
-    // MARK: - Public Properties
-
-    open weak var delegate: MessageLabelDelegate?
+import UIKit
+extension NSTextStorage {
+    
+    var attributedString: NSAttributedString {
+        let range = NSRange(location: 0, length: mutableString.length)
+        return attributedSubstring(from: range)
+    }
+    
+    var mutableAttributedString: NSMutableAttributedString {
+        return NSMutableAttributedString(attributedString: attributedString)
+    }
+    
+    func addAttribute(key: NSAttributedStringKey, value: AnyObject, range: NSRange? = nil) {
+        let range = range ?? NSRange(location: 0, length: length)
+        addAttribute(key, value: value, range: range)
+    }
+    
+}
 
-    open var enabledDetectors: [DetectorType] = [.phoneNumber, .address, .date, .url]
+open class MessageLabel: UIView, UIGestureRecognizerDelegate {
 
-    open override var attributedText: NSAttributedString? {
-        didSet {
-            guard attributedText != oldValue else { return }
-            setTextStorage()
+    // MARK: - Properties [Public]
+    
+    public weak var delegate: MessageLabelDelegate?
+    
+    public var attributedText: NSAttributedString? {
+        get {
+            return textStorage.attributedString
+        }
+        set {
+            let string = newValue ?? NSAttributedString(string: "")
+            textStorage.setAttributedString(string)
+            
+            setStorageByParsing(text: string) {
+                if !self.isConfiguring { self.setNeedsDisplay() }
+            }
+            
+            if !isConfiguring { setNeedsDisplay() }
         }
     }
-
-    open override var text: String? {
-        didSet {
-            guard text != oldValue else { return }
-            setTextStorage()
+    
+    public var text: String? {
+        get {
+            return textStorage.string
+        }
+        set {
+            let attributedString = NSAttributedString(string: newValue ?? "")
+            let string = addLocalAttributes(to: attributedString)
+            textStorage.setAttributedString(string)
+            
+            setStorageByParsing(text: string) {
+                if !self.isConfiguring { self.setNeedsDisplay() }
+            }
+            
+            if !isConfiguring { setNeedsDisplay() }
```

Comment:
Why are we calling setNeedsDisplay twice here?

Reply:
The one inside the block is `async`. We need some refactoring around this for sure. I will probably add a flag to the configure method: `configure(parseText: Bool)` or something

Reply:
👍